Title
Author: Kerria Burns
Affiliation: UVM
E-mail contact: Kerria.Burns@uvm.edu
Start Date: 2020-01-13
End Date: 2020-05-08
Project Descriptions: Class notes for ecological genomics
Table of Contents:
Entry 1: 2020-01-13, Monday First day of class
Entry 2: 2020-01-14, Tuesday
Entry 3: 2020-01-15, Wednesday Second day of class Methods info update blitz and computer set-up
Entry 4: 2020-01-16, Thursday
Entry 5: 2020-01-17, Friday
Entry 6: 2020-01-20, Monday
Entry 7: 2020-01-21, Tuesday
Entry 8: 2020-01-22, Wednesday Additional sequencing info updates
Entry 9: 2020-01-23, Thursday
Entry 10: 2020-01-24, Friday
Entry 11: 2020-01-27, Monday Population genomics and paper discussion
Entry 12: 2020-01-28, Tuesday
Entry 13: 2020-01-29, Wednesday
Entry 14: 2020-01-30, Thursday
Entry 15: 2020-01-31, Friday
Entry 16: 2020-02-03, Monday
Entry 17: 2020-02-04, Tuesday
Entry 18: 2020-02-05, Wednesday
Entry 19: 2020-02-06, Thursday
Entry 20: 2020-02-07, Friday
Entry 21: 2020-02-10, Monday
Entry 22: 2020-02-11, Tuesday
Entry 23: 2020-02-12, Wednesday
Entry 24: 2020-02-13, Thursday
Entry 25: 2020-02-14, Friday
Entry 26: 2020-02-17, Monday
Entry 27: 2020-02-18, Tuesday
Entry 28: 2020-02-19, Wednesday
Entry 29: 2020-02-20, Thursday
Entry 30: 2020-02-21, Friday
Entry 31: 2020-02-24, Monday
Entry 32: 2020-02-25, Tuesday
Entry 33: 2020-02-26, Wednesday
Entry 34: 2020-02-27, Thursday
Entry 35: 2020-02-28, Friday
Entry 36: 2020-03-02, Monday
Entry 37: 2020-03-03, Tuesday
Entry 38: 2020-03-04, Wednesday
Entry 39: 2020-03-05, Thursday
Entry 40: 2020-03-06, Friday
Entry 41: 2020-03-09, Monday
Entry 42: 2020-03-10, Tuesday
Entry 43: 2020-03-11, Wednesday
Entry 44: 2020-03-12, Thursday
Entry 45: 2020-03-13, Friday
Entry 46: 2020-03-16, Monday
Entry 47: 2020-03-17, Tuesday
Entry 48: 2020-03-18, Wednesday
Entry 49: 2020-03-19, Thursday
Entry 50: 2020-03-20, Friday
Entry 51: 2020-03-23, Monday
Entry 52: 2020-03-24, Tuesday
Entry 53: 2020-03-25, Wednesday
Entry 54: 2020-03-26, Thursday
Entry 55: 2020-03-27, Friday
Entry 56: 2020-03-30, Monday
Entry 57: 2020-03-31, Tuesday
Entry 58: 2020-04-01, Wednesday
Entry 59: 2020-04-02, Thursday
Entry 60: 2020-04-03, Friday
Entry 61: 2020-04-06, Monday
Entry 62: 2020-04-07, Tuesday
Entry 63: 2020-04-08, Wednesday
Entry 64: 2020-04-09, Thursday
Entry 65: 2020-04-10, Friday
Entry 66: 2020-04-13, Monday
Entry 67: 2020-04-14, Tuesday
Entry 68: 2020-04-15, Wednesday
Entry 69: 2020-04-16, Thursday
Entry 70: 2020-04-17, Friday
Entry 71: 2020-04-20, Monday
Entry 72: 2020-04-21, Tuesday
Entry 73: 2020-04-22, Wednesday
Entry 74: 2020-04-23, Thursday
Entry 75: 2020-04-24, Friday
Entry 76: 2020-04-27, Monday
Entry 77: 2020-04-28, Tuesday
Entry 78: 2020-04-29, Wednesday
Entry 79: 2020-04-30, Thursday
Entry 80: 2020-05-01, Friday
Entry 81: 2020-05-04, Monday
Entry 82: 2020-05-05, Tuesday
Entry 83: 2020-05-06, Wednesday
Entry 84: 2020-05-07, Thursday
Entry 85: 2020-05-08, Friday
Entry 1: 2020-01-13, Monday.
Introductions
What is ecological genomics?
Species interactions at the genomic level
Study of genes and relation to organisms in their environment
In the context of evolution
Overlapping of fields
Popgen: 1920/30s Wright, Fisher, Haldane

Mathematical theory: based on a theoretical unchanging population
Fst population level differentiation
Selection: thinking in terms of the relative fitness of different genotypes
Molecular evolution: Kimura Tajima

How does the DNA reveal the history of an organism/population/species
Can look at pos/neg selection by comparing nonsynonymous to synonymous changes in the DNA
Can look at the nucleotide diversity, processes of evolution can influence diversity stats
Merge these two fields with ecology and NGS -> ecological genomics

Class structure
Mondays: mini lectures/info updates and journal review Wednesdays: Computer lab

Entry 2: 2020-01-14, Tuesday.
Entry 3: 2020-01-15, Wednesday.
How to pick a platform/method
Question
Organisms / sample (tissue type, environmental)
Is there a reference genome?
Ploidy
DNA/RNA/?
Different library prep methods
Whole genome sequencing
RNAseq
Exome capture
GBS/Rad
Amplicon
Bisulfite/ATAC
Sequencing platforms
Illumina
Pacbio/nanopore
Info updates
Chege: Illumina
Video on Illumina platform

Library prep
Cut up DNA/RNA into pieces and attach adapters
Cluster generation
Bind fragments of DNA to lawn of complementary sequences to adapters
Through PCR process our fragments are magnified, to facilitate sequencing
Sequencing by synthesis
Add fluorescent dNTPs
As each base is read an image is stored
Data analysis
Why do we use this method?

MiSeq: 30mill bp /run
HiSeq: 3 bill /run
NovaSeq: 13 bill / run
Sanger: 400 /run
Csenge: PacBio/Nanopore
Two different approaches: SMRT or Synthetic
PacBio and Nanopore both use SMRT method
PacBio
Utilize special flow cell with a bunch of small wells
Polymerase is fixed at the bottom of the well and the DNA passes across the wells
Fluorescence is recorded as each nucleotide is incorporated
Nanopore
directly detects sequence of DNA
Sequence moves through a pore and instead of looking at fluorescence of each base ... has unique sequence signal
These methods are less accurate than short read sequencing and more expensive
Can help with repetitive regions of DNA, get rid of potential gaps, helpful to identify alternative splicing
Erika: RNAseq
Whole transcriptome shotgun sequencing
Uses high throughput NGS technology to characterize gene expression patterns and to identify novel transcripts/molecular markers of interest
Can incorporate into your experimental framework
Advantage: gives functionally relevant information about what is being expressed at that time point. Can also allow you to detect differences among populations (both gene expression and SNP differences).
Limitations: some inherent error, RNA is a less stable molecule can introduce error/bias
Need to be mindful of expression level differences among tissue types, etc.
Workflow: Sample collection (freeze/RNAlater), isolate RNA, quality assessment of RNA (gel or bioanalyzer or qubit), rRNA depletion (selecting for molecules with poly-A tail to enrich for mRNA), fragmentation, cDNA synthesis, adding adapters (amplification element, primary sequencing site, barcode for multiplexing), amplification using PCR, sequence on platform of choice
Zoe: GBS/RADseq
Use restriction sites to fragment DNA
Ligate adapters to each of your fragments with a barcode
Workflow: Extract genomic DNA from your samples, digest with a restriction enzyme, ligate adapters, multiplex individuals, clean up fragments, PCR, check fragment size, send for sequencing
Limitations: fragment bias for what gets amplified
Ben: Bisulfite/ATAC seq
Bisulfite: used to identify methylated cytosine, which typically occurs in CG islands, typical of non-coding DNA or in promotors upstream of genes that aren't constantly expressed.
Workflow: Isolate DNA, treat with sodium bisulfite (converts meytholated Cs into Us), PCR, Sequence samples and compare treated vs untreated samples (these CG islands will differ in their sequence) so this allows you to identify methylated vs non-methylated regions of the genome
CHIPseq: acts at the level of histone modification, chromatin immunoprecipitation
Workflow: introduce a fixative to the DNA, fragment the DNA, introduce an antibody that is specific to your protein of interest which allows your targeted fragments to be precipitated out, decouple sequence and protein, and then you sequence
ATACseq: looks at chromatin state
Workflow: introduce transposons that fragment open chromatin, transposons add adapters to fragments, sequence fragments
HiC seq: identifies all regions where the DNA strand is interacting with itself and sequences them
Entry 4: 2020-01-16, Thursday.
Entry 5: 2020-01-17, Friday.
Entry 6: 2020-01-20, Monday.
Entry 7: 2020-01-21, Tuesday.
Entry 8: 2020-01-22, Wednesday.
Jorge: Amplicon Seq
Uses:

Genome targeting
Detection of hot-spots of mutation
Gene fusion
SNPs
Metagenomics
Workflow:

Sample selection
DNA or RNA extraction
Select amplicon that you will sequence: ITS, CO1, 16S
Library prep
PCR using primers to focus on amplicon of choice
Optional purification step
2nd PCR to attach barcode
Size selection purification to isolate your amplicon
Pool samples
Sequence
Bioinformatic procedure
Quality contreol
trimming
demultiplex
generate your OTUs (operational taxonomic unit)
Alison: Exon capture sequencing
Exon: protein coding region of the gene Exome: portion of the genome that codes for genes

Prevalent in biomedical field and in evolutionary biology
More targeted approach than WGS Biomedical uses:
disease causing variants are more likely to be found in the exome
especially interested for rare diseases
EcoEvo Uses:

Genetic mapping of phenotypic traits
Detect selection (particularly helpful for non-model organisms)
Used in phylogenetic reconstruction
Can also be used to analyze ancient DNA
Workflow:

Sample collection & extraction
Fragmentation of DNA
Enriching library using probes
PCR: need to have primers designed
Denovo: create new probes, need some knowledge of transcriptome/RADseq/WGS
Divergent annotated genome: probes based on relatively close relative, requires annotate genomes
DNA purification
Quality control
Entry 9: 2020-01-23, Thursday.
Entry 10: 2020-01-24, Friday.
Entry 11: 2020-01-27, Monday.
#####Steve lecture on population genomics
variation across the genome
  statistally capture variation with statistics
    number of SNPs dependent on location in chromosome (LD)
      selection acts locally across chromosome
      find outliers
    demographic processes act globally
      mean and distribution
    demographic history
      Ne: idealized Wright-Fisher population
      -migration
      -genetic populations (demes)
        divergence times between
      -bottleneck, expansion, splitting
      -coalescent theory
        -branch of population genetics
        -probability that two sampled allele copies share a common ancestor
          -probability that two haplotypes originated from the same genotype
      -coalescence speeds up with lower population size
        -1/2Ne
        -due to increased effects of genetic drift in smaller populations
        -estimation migration through whether there are allele copies with much deeper coalescence
          -allele copy originated far back in common ancestor but migrated
            -common ancestor not present since population was unified
        -presence of mutation distinct from geneaology
          -fuzzy image
        -almost all alleles are rare
        -if population is growing, see a lot of rare mutations
          -frequency of rare alleles is very high, frequency of common alleles is low
            -rate at which mutations enter population is proportional to effective population size times mutation rate
              -theta=Ne*u
        -bottlenecked populations lose rare alleles and retain common alleles
      -effect of selection on allele frequencies depends on type
        -focus here is positive selection
      -pi=nucleotide diversity
        -average pairwise differences
          -sample two individuals: how many SNPs do they differ by in the chromosome?
        -LD measures extent of correlation
          -hitchhiking (not under seleciton themselves, just along for the ride with selected sites)
      -selective sweep has same signature as population expansion
      -Taj D very negative in case of selective sweep or population expansion
        -to distinguish selection, have to pick out local signals
Entry 12: 2020-01-28, Tuesday.
Entry 13: 2020-01-29, Wednesday.

URL: [what to hyperlink](URL)

embed code: /// cd ~/mydata ///

This is where the class data are `/data/project.data`

Phred score (probability that reader misread the nucleotide on the flow cell (could still be the wrong base)): p=10^(-Q/10)
e.g., p=10^(-38/10)=0.0038 -> a good score

XFS
run fastqc on trimmed sequences
did it clean things up? did quality scores improve? especially at the beginning and end?
just on cleaned, paired reads.

look back at cd (ll) /data/project_data/RS_ExomeSeq/fastq/edge_fastq/pairedcleanreads/
```
#!/bin/bash

cd ~/Ecological_Genomics/myresults

#creating new dir to store results
mkdir fastqc

for file in /data/project_data/RS_ExomeSeq/fastq/edge_fastq/XFS*fastq.gz

do

fastqc ${file} -o fastqc/

done
```
Entry 14: 2020-01-30, Thursday.
Entry 15: 2020-01-31, Friday.
Entry 16: 2020-02-03, Monday.
Entry 17: 2020-02-04, Tuesday.
Entry 18: 2020-02-05, Wednesday.

#!/bin/bash

#This script will run the read mapping using "bam"

ref="/data/project_data/RS_ExomeSeq/ReferenceGenomes/Pabies1.0-genome_reduced.fa"

#Write a loop to map each individual within my population.

for forward in ${input}*_R1.cl.pd.fq

do
  reverse=${forward/_R1.cl.pd.fq/_R2.cl.pd.fq}
  f=${forward/_R1.cl.pd.fq/}
  name=`basename ${f}`
  bwa mem -t 1 -M ${ref} ${forward} ${reverse} > ${output}/BWA/${name}.sam
done


#!/bin/bash   
 
cd /data/project_data/RS_ExomeSeq/fastq/edge_fastq  
 
for R1 in XFS*R1_fastq.gz  

do 
 
	R2=${R1/_R1_fastq.gz/_R2_fastq.gz}
	f=${R1/_R1_fastq.gz/}
	name=`basename ${f}`

	java -classpath /data/popgen/Trimmomatic-0.33/trimmomatic-0.33.jar org.usadellab.trimmomatic.TrimmomaticPE \
        -threads 1 \
        -phred33 \
         "$R1" \
         "$R2" \
         /data/project_data/RS_ExomeSeq/fastq/edge_fastq/pairedcleanreads/${name}_R1.cl.pd.fq \
         /data/project_data/RS_ExomeSeq/fastq/edge_fastq/unpairedcleanreads/${name}_R1.cl.un.fq \
         /data/project_data/RS_ExomeSeq/fastq/edge_fastq/pairedcleanreads/${name}_R2.cl.pd.fq \
         /data/project_data/RS_ExomeSeq/fastq/edge_fastq/unpairedcleanreads/${name}_R2.cl.un.fq \
        ILLUMINACLIP:/data/popgen/Trimmomatic-0.33/adapters/TruSeq3-PE.fa:2:30:10 \
        LEADING:20 \
        TRAILING:20 \
        SLIDINGWINDOW:6:20 \
        MINLEN:35 
 
done 
Entry 19: 2020-02-06, Thursday.
Entry 20: 2020-02-07, Friday.
Entry 21: 2020-02-10, Monday.
Entry 22: 2020-02-11, Tuesday.
Entry 23: 2020-02-12, Wednesday.
#!/bin/bash

#We'll use this wrapper to run our different scripts.

#Path to my repo

myrepo="/users/k/b/kburns/EcologicalGenomics/"

#My population
mypop="XFS"

#Directory to our cleaned and paired reads
input="/data/project_data/RS_ExomeSeq/fastq/edge_fastq/pairedcleanreads/${mypop}"

#Directory to store the outputs on our mapping
output="/data/project_data/RS_ExomeSeq/mapping"


#Run mapping.sh
#source ./mapping.sh

#Run the post-processing steps
source ./process_bam.sh


#for f in ${output}/BWA/${mypop}*.sam

#do

#  out=${f/.sam/}
 # sambamba-0.7.1-linux-static view -S --format=bam ${f} -o ${out}.bam
  #samtools sort ${out}.bam -o ${out}.sorted.bam

#done

#Now, let's remove the PCR duplicates
#for file in ${output}/BWA/${mypop}*.sorted.bam

#do
 # f=${file/.sorted.bam/}
  #sambamba-0.7.1-linux-static markdup -r -t 1 ${file} ${f}.sorted.rmdup
#done

#Now, to finish we'll index our files
for file in ${output}/BWA/${mypop}*.sorted.rmdup.bam

do
  samtools index ${file}
done
To account for low read scores, use genotype likelihood to account for uncertainty.
Entry 26: 2020-02-17, Monday.
Entry 27: 2020-02-18, Tuesday.
Entry 28: 2020-02-19, Wednesday.
ANGSD script:
myrepo="/users/k/b/kburns/EcologicalGenomics"

mkdir ${myrepo}/myresults/ANGSD

output="${myrepo}/myresults/ANGSD"

mypop="XFS"

ls /data/project_data/RS_ExomeSeq/mapping/BWA/${mypop}_*sorted.rm*.bam >${output}/${mypop}_bam.list

REF="/data/project_data/RS_ExomeSeq/ReferenceGenomes/Pabies1.0-genome_reduced.fa"

# Estimating GL's and allele frequencies for all sites with ANGSD

ANGSD -b ${output}/${mypop}_bam.list \
-ref ${REF} -anc ${REF} \
-out ${output}/${mypop}_allsites \
-nThreads 1 \
-remove_bads 1 \
-C 50 \
-baq 1 \
-minMapQ 20 \
-minQ 20 \
-setMinDepth 3 \
-minInd 2 \
-setMinDepthInd 1 \
-setMaxDepthInd 17 \
-skipTriallelic 1 \
-GL 1 \
-doCounts 1 \
-doMajorMinor 1 \
-doMaf 1 \
-doSaf 1 \
-doHWE 1 \
# -SNP_pval 1e-6



ANGSD -b ${output}/${mypop}_bam.list \
-ref ${REF}  \
-out ${output}/${mypop}_folded_allsites \
-nThreads 1 \
-remove_bads 1 \
-C 50 \
-baq 1 \
-minMapQ 20 \
-minQ 20 \
-setMinDepth 3 \
-minInd 2 \
-setMinDepthInd 1 \
-setMaxDepthInd 17 \
-skipTriallelic 1 \
-GL 1 \
-doCounts 1 \
-doMajorMinor 1 \
-doMaf 1 \
-doSaf 1 \
-fold 1 




Estimating rough SFS
.sfs
Estimate SFS using .sfs as a prior
doTheta=4Neu
theta_stats





setwd("~/Documents/GitHub/EcologicalGenomics/myresults")

list.files()

SFS <- scan("XFS_outFold.sfs")

sumSFS <- sum(SFS)


pctPoly = 100*(1-(SFS[1]/sumSFS))

plotSFS <- SFS[-c(1,length(SFS))]

barplot(plotSFS)

div <- read.table("XFS_folded_allsites.thetas.idx.pestPG")

colnames(div)=c("window","chrname","wincenter","tW","tP","tF","tH","tL","tajD","fulif","fuliD","fayH","zengsE","numSites")

div$tWpersite=div$tW/div$numSites
div$tPpersite=div$tP/div$numSites

hist(div$tWpersite-div$tPpersite)

pdf("XFS_diversity_stats.pdf")
par(mfrow=c(2,2))
hist(div$tWpersite, col="gray",xlab="Theta-W",main="")
hist(div$tPpersite, col="gray",xlab="Theta-Pi",main="")
hist(div$tajD, col="gray",xlab="Tajima's D",main="")
barplot(plotSFS)
dev.off()

summary(div)

head(div)

Entry 29: 2020-02-20, Thursday.
Entry 30: 2020-02-21, Friday.
Entry 31: 2020-02-24, Monday.
Entry 32: 2020-02-25, Tuesday.
Entry 33: 2020-02-26, Wednesday.
Entry 34: 2020-02-27, Thursday.
Entry 35: 2020-02-28, Friday.
Entry 36: 2020-03-02, Monday.
Entry 37: 2020-03-03, Tuesday.
Entry 38: 2020-03-04, Wednesday.
Entry 39: 2020-03-05, Thursday.
Entry 40: 2020-03-06, Friday.
Entry 41: 2020-03-09, Monday.
Entry 42: 2020-03-10, Tuesday.
Entry 43: 2020-03-11, Wednesday.
Entry 44: 2020-03-12, Thursday.
Entry 45: 2020-03-13, Friday.
Entry 46: 2020-03-16, Monday.
Entry 47: 2020-03-17, Tuesday.
Entry 48: 2020-03-18, Wednesday.
Entry 49: 2020-03-19, Thursday.
Entry 50: 2020-03-20, Friday.
Entry 51: 2020-03-23, Monday.
Entry 52: 2020-03-24, Tuesday.
Entry 53: 2020-03-25, Wednesday.
Entry 54: 2020-03-26, Thursday.
Entry 55: 2020-03-27, Friday.
Entry 56: 2020-03-30, Monday.
Entry 57: 2020-03-31, Tuesday.
Entry 58: 2020-04-01, Wednesday.
Entry 59: 2020-04-02, Thursday.
Entry 60: 2020-04-03, Friday.
Entry 61: 2020-04-06, Monday.
Entry 62: 2020-04-07, Tuesday.
Entry 63: 2020-04-08, Wednesday.
Entry 64: 2020-04-09, Thursday.
Entry 65: 2020-04-10, Friday.
Entry 66: 2020-04-13, Monday.
Entry 67: 2020-04-14, Tuesday.
Entry 68: 2020-04-15, Wednesday.
Entry 69: 2020-04-16, Thursday.
Entry 70: 2020-04-17, Friday.
Entry 71: 2020-04-20, Monday.
Entry 72: 2020-04-21, Tuesday.
Entry 73: 2020-04-22, Wednesday.
Entry 74: 2020-04-23, Thursday.
Entry 75: 2020-04-24, Friday.
Entry 76: 2020-04-27, Monday.
Entry 77: 2020-04-28, Tuesday.
Entry 78: 2020-04-29, Wednesday.
Entry 79: 2020-04-30, Thursday.
Entry 80: 2020-05-01, Friday.
Entry 81: 2020-05-04, Monday.
Entry 82: 2020-05-05, Tuesday.
Entry 83: 2020-05-06, Wednesday.
Entry 84: 2020-05-07, Thursday.
Entry 85: 2020-05-08, Friday.
